### DISCUSSION TITLE
"unsupported type" using POST with JSON data

### DISCUSSION THREAD

**Discussion by 3goats** [rqlite/rqlite#1161] [General]
Sorry for the question. I'm a little confused and this maybe stupid.

If I start the Rqlite shell and do this: 

```
create table user(id integer primary key, name text, interests json);
sqlite> insert into user values(null, "John", '{"likes": ["skating", "reading", "swimming"], "dislikes": ["cooking"]}');`
1 row affected
```

All seem to work fine. 

If I try to do this via cURL: 

```
curl --location --request POST 'localhost:4001/db/execute?pretty&timings' \
--header 'Content-Type: application/json' \
--data-raw '[
   ["INSERT INTO user(name, interests) VALUES(?, ?)", "fiona", {"likes": ["skating", "reading", "swimming"], "dislikes": ["cooking"]}]
]'
unsupported type
```

It doesn't work. However, if I do this: 

```
curl -XPOST 'localhost:4001/db/execute?pretty&timings' -H "Content-Type: application/json" -d '[
   ["INSERT INTO user(name, interests) VALUES(?, ?)", "fiona", "{\"likes\": [\"skating\", \"reading\", \"swimming\"], \"dislikes\": [\"cooking\"]}"]
]'
{
    "results": [
        {
            "last_insert_id": 23,
            "rows_affected": 1,
            "time": 0.000048291
        }
    ],
    "time": 0.011988834
}
```

It works. 

Is there a way to avoid  having to escape everything and send plain JSON structures via the HTTP API like I do using the rqlite shell  ?

Based on my schema the "interests" column has a type of JSON, so should this not work ? 

```
pragma table_info('user');
+-----+-----------+---------+---------+------------+----+
| cid | name      | type    | notnull | dflt_value | pk |
+-----+-----------+---------+---------+------------+----+
| 0   | id        | INTEGER | 0       | <nil>      | 1  |
+-----+-----------+---------+---------+------------+----+
| 1   | name      | TEXT    | 0       | <nil>      | 0  |
+-----+-----------+---------+---------+------------+----+
| 2   | interests | json    | 0       | <nil>      | 0  |
+-----+-----------+---------+---------+------------+----+
```

**Comment by otoolep** [comment]
>Is there a way to avoid having to escape everything and send plain JSON structures via the HTTP API like I do using the rqlite shell ?

There is no way to avoid when using the bash shell. The rqlite shell is doing the escaping for you:

https://github.com/rqlite/rqlite/blob/master/cmd/rqlite/execute.go#L38

Does that answer your question? It's a limitation of bash, not something specific to rqlite.

**Reply by 3goats** [reply]
Ok that explains it - wasn't sure if it was expected behaviour or not. I'll work around it as you say using python or go. Thanks for the reply

**Comment by otoolep** [comment]
You could also make it easier for yourself, perhaps, by writing a small tool that you can call directly from bash -- something built using the Python or Go client libraries.

**Comment by 3goats** [comment]
So an interesting thing is I tried this using Postman: 

```
POST [
    ["INSERT INTO user(name, interests) VALUES(:name, :interests)", {"name": "fiona", "interests": {
	"likes": ["skating", "reading", "swimming"],
	"dislikes": ["cooking"]
}}]
] 
```

And I get the same: `unsupported type` error

**Comment by otoolep** [comment]
Without knowing what Postman is doing under the covers, it's hard to offer to an opinion.
